### PR TITLE
EQL: Add integration tests harness to test EQL feature parity with original implementation

### PR DESCRIPTION
**Pull request by aleksmaus**
The tests use the original test queries from
https://github.com/endgameinc/eql/blob/master/eql/etc/test_queries.toml
for EQL implementation correctness validation.
The file test_queries_unsupported.toml serves as a "blacklist" for the
queries that we do not support. Currently all of the queries are
blacklisted. Over the time the expectation is to eventually have an
empty "blacklist" when all of the queries are fully supported.

The tests use the original test vector from
https://raw.githubusercontent.com/endgameinc/eql/master/eql/etc/test_data.json
that was translated to ECS format that matches the latest mapping being
used for Endgame platform event streaming and is loaded from endgame.dat
file. The endgame.json file contains the matching index
mappings/setting.

Only one EQL spec response is stubbed for now to match the expected
output for that spec query. This part would need some tweaking after EQL is
fully wired.

The input .toml file is parsed by hand for now, which is sufficient for
our purposes and avoids introducing another dependency just for this
particular test case.

Related to https://github.com/elastic/elasticsearch/issues/49581

@imotov could you take a look first before I promote this from draft PR?
Thanks!

### CODE REVIEW

**Comment by aleksmaus** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix-unix

File: x-pack/plugin/eql/src/test/resources/endgame.dat
Comment (rw-access) [inline x-pack/plugin/eql/src/test/resources/endgame.dat]: two q's
- why the `.dat` extension?
- I don't think this needs to be in ECS, other than the event.category field. Otherwise all of the queries need to be rewritten

File: x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml
Comment (rw-access) [inline x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml]: this one is supported

Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml]: yeah, but not hooked up to the query engine yet. Once it is, we can validate the results and update the .toml files.

File: x-pack/plugin/eql/src/test/resources/endgame.dat
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/resources/endgame.dat]: 1. it's not quite a valid json, it's a collection of jsons. open for suggestions on the file extension.
2. wanted to use the exact same ECS data format that is currently established for events streaming from SMP, so we could potentially flash out the details.

File: x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml]: expecting the test_queries_unsupported.toml file to become empty and eventually dropped when we reach the feature parity

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java
Comment (imotov) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java]: I don't think we need this anymore.

**Review by imotov — COMMENTED** [review]
I think @costin and @nik9000 might be better reviews for this since they had more experience with sql-spec tests.

**Comment by aleksmaus** [comment]
@imotov,  I talked to @costin before and got few suggestions. Basically, wanted to keep the original test spec from EQL unchanged, so we can easily follow any original EQL implementation changes by copying the toml file over into our test harness. Also tried to avoid adding any toml library dependency, since the format of the file is pretty simple to parse as is without the need for a full toml parser implementation yet, especially for testing.

**Review by nik9000 — COMMENTED** [review]
I'm certainly ok reusing the tests from other spots so the toml is fine by me. I think it'd be better to pull in an apache licensed toml parser as a test dependency for this project rather than build it by hand. Test dependencies are generally not so big a deal, especially when we isolate them into a specific sub-project.

The fancy CSV tests that SQL has have the nice property of being able to include them in the docs because they *look* just like output from the command line sql util. toml doesn't have that property, but that seems like a small price to pay for being able to reuse tests from a mature project.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java]: Either we use `enabled` or this method. On one hand `enabled` checks the property presence statically but this method tries to do it dynamic when a class is instantiated.
There's there are multiple pieces of `isSnapshot` in this class that need simplification.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java]: Same as above, formatting on non-modified lines is noise.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (costin) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: nit: the preference in general is not to negate but instead use `== false` for clarity.

File: x-pack/plugin/sql/build.gradle
Comment (costin) [inline x-pack/plugin/sql/build.gradle]: Why?

**Review by costin — COMMENTED** [review]
Left some comments.

I'm on the fence with the TOML reader. On one hand the parser is already written and we can customize it appropriately - on the other hand using an existing library would reduce (theoretically) the maintenance burdon.

I don't know enough about TOML to tell how much standard/non-standard it can get.
Regardless, reusing the test queries and results in the docs is something that's valuable to pursue in the future.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java]: Just making it easier to read, also because it just was just brought to my attention that some of the code that I had some formatting issues.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java]: consistent with other plugins. can undo

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java]: tried to follow the autoscaling plugin pattern, that I was told should be an example of handling the feature flag.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java]: this is the formatting misalignment that I missed in my previous PR, now trying to fix it

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: will update

File: x-pack/plugin/sql/build.gradle
Comment (aleksmaus) [reply x-pack/plugin/sql/build.gradle]: will remove

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java]: Import static for `IMMEDIATE`?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java]: Any reason why not making `query` private inside `EqlSpec` and provide a `query()` accessor method?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: Will this method be used in the future? (atm it isn't)

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: Imo, it would help to see a method comment containing sample line(s) these `read*` methods are expected to handle.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: Why are all these `read*` methods trimming the values they are reading? For example, `readTags` gets its values from `readArray` which is already trimming the values returned as array...
Are all these `trim()` calls really needed or they are added just "to be safe"?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: I don't think this initialization is necessary. A simple `String delim;` should be enough.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: Single `try()` line.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java
Comment (nik9000) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java]: I think this test should use `@ParametersFactory` so each example becomes a test that we can address individually.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: There is a class `org.elasticsearch.common.Strings` that has the same method and I think we prefer that one.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: Same here about `Strings` class.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java]: Can you remove this, please?

**Review by astefan — COMMENTED** [review]
Good stuff. Left some comments.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java]: I would not sweat about this particular code, it will be cleaner once we remove the feature flag.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: EqlSpecLoader.load is used here 
https://github.com/elastic/elasticsearch/pull/52248/files#diff-663165787e62c937a0aa8ac3d8511d6dR63

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: in this particular case the trimming is done for the case where the value in the brackets prefixed/suffixed with spaces, for example: " foo bar "

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: IDE was complaining on unitialized var delim otherwise

File: x-pack/plugin/eql/src/test/resources/endgame.dat
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/resources/endgame.dat]: Updated to .ndjson

**Comment by aleksmaus** [comment]
@nik9000 replacing the manual toml parsing with io.ous:jtoml:2.0.0.
The other couple of libraries that I looked at depend on antlr4  4.7.1 which breaks compatibility with existing 4.5.3.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: this is no longer relevant, replacing with toml library

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/EqlTestUtils.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/EqlTestUtils.java]: No longer used indeed, will remove

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/AbstractEqlIntegTestCase.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/AbstractEqlIntegTestCase.java]: Probably. Was following AbstractSqlIntegTestCase as example in this case.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java]: No longer relevant, replaced with the toml parser library

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java]: Seems like this is consistent with 100+ other places 
Screen Shot 2020-02-18 at 2 00 58 PM

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java]: Less typing for the test code. Will convert to accessors.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: No longer relevant, replaced with toml parser

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: No longer relevant, replaced with toml parser.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: No longer relevant, replaced with toml parser.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java]: Updated

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java
Comment (aleksmaus) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java]: Sure. It was initially useful to have while we were iterating on the response shape.

File: x-pack/plugin/eql/src/test/resources/endgame.json
Comment (rw-access) [inline x-pack/plugin/eql/src/test/resources/endgame.json]: nit: a more descriptive filename. i.e. endgame_mappings.json

File: x-pack/plugin/eql/src/test/resources/endgame.ndjson
Comment (rw-access) [inline x-pack/plugin/eql/src/test/resources/endgame.ndjson]: the mappings are consistent with the data, but the queries are not. EQL is independent of ECS, so I think it's okay to leave the data in the same schema that it was in

File: x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml
Comment (rw-access) [reply x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml]: oh, I didn't realize that. in that case, this file is missing the first query from test_queries.toml. i just looked at the top and saw the discrepancy and reached the wrong conclusion
```toml
[[queries]]
query = 'process where serial_event_id = 1'
expected_event_ids  = [1]
```

File: x-pack/plugin/eql/src/test/resources/endgame.ndjson
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/resources/endgame.ndjson]: using this format mostly in order to validate and work through any possible issues with covering our particular use case

Comment (rw-access) [reply x-pack/plugin/eql/src/test/resources/endgame.ndjson]: that means we have to manually rewrite all of the queries though, which is likely more error prone.

Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/resources/endgame.ndjson]: Discussed with @rw-access on Slack. 
Originally I was under impression that we will keep the queries/rules the same and use some internal mapping to adopt them for ECS.  I looks like in the first iteration this will not be the case and the queries/rules will be rewritten for ECS.
I will update this PR to use the same flat json data from the original EQL tests and remove the ECS mapping template creation for now.

**Comment by aleksmaus** [comment]
Updated PR with the latest recommendations from @rw-access, also including @nik9000 suggestion to use @ParametersFactory for the test harness.

**Comment by aleksmaus** [comment]
Added a small adjustment after merging the latest master changes that included @imotov HL client test against the currently hardcoded EQL response. I changed the shape of the response for the first of the EQL validation harness test queries expected response, while Igor was working on HL client.

File: x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml]: Added the comment into the file per @rw-access recommendation.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by aleksmaus** [comment]
Had to merge master again and resolve conflicts. Will merge once it clears CI build.

**Pull request merged**
